### PR TITLE
Closes #396: Write university section aware hash bucket

### LINKED ISSUE #396: Write university section aware doc and org hashers

### PR DESCRIPTION
**Pull request by madryk**

### CODE REVIEW

**Comment by mpol** [comment]
Unit tests fail:
```
Tests in error: 
  AffiliationMainSectionBucketHasherTest.hash:55 » IllegalArgument
  AffiliationMainSectionBucketHasherTest.hash_fallback_section:84 » IllegalArgument
  OrganizationMainSectionBucketHasherTest.hash:55 » IllegalArgument
  OrganizationMainSectionBucketHasherTest.hash_fallback_section:84 » IllegalArgument
```

**Comment by madryk** [comment]
I've corrected tests

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/bucket/AffiliationMainSectionBucketHasher.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import eu.dnetlib.iis.wf.affmatching.orgsection.OrganizationSection.OrgSectionType;
+import eu.dnetlib.iis.wf.affmatching.orgsection.OrganizationSectionsSplitter;
+
+/**
+ * An implementation of {@link BucketHasher} that hashes {@link AffMatchOrganization}.
+ * Hash is generated from 'main' (most significant) section of {@link AffMatchOrganization#getName()}.
+ * 
+ * @author madryk
+ */
+public class AffiliationMainSectionBucketHasher implements BucketHasher<AffMatchAffiliation> {
```
Comment (mpol) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/bucket/AffiliationMainSectionBucketHasher.java]: Code duplication with `OrganizationMainSectionBucketHasher`. Probably good idea to extract common code to some `MainSectionBucketHasher`.

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/bucket/OrganizationSectionHasher.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * Class that generates hash of the given {@link OrganizationSection}.
+ * 
+ * @author madryk
+ */
+public class OrganizationSectionHasher implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final int SECTION_NUMBER_OF_LETTERS = 3;
```
Comment (mpol) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/bucket/OrganizationSectionHasher.java]: I suppose this number depends on the number of letters required to disambiguate different section types? Then maybe it should be a constant in `OrganizationSection`?

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/orgsection/OrganizationSectionsSplitter.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    
+    private int findWordStartingWithAny(String[] words, String ... wordStart) {
+        
+        for (int i=0; i<words.length; ++i) {
+            for (int j=0; j<wordStart.length; ++j) {
+                
+                if (words[i].startsWith(wordStart[j])) {
+                    return i;
+                }
+            }
```
Comment (mpol) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/orgsection/OrganizationSectionsSplitter.java]: The inner loop is `StringUtils.startsWithAny`

**Pull request closed without merging**